datafusion/functions/src/encoding/inner.rs

```diff
@@ -232,324 +210,286 @@ impl ScalarUDFImpl for DecodeFunc {
-                }
-                ScalarValue::Binary(a) => Ok(
-                    encoding.encode_scalar(a.as_ref().map(|v: &Vec<u8>| v.as_slice()))
-                ),
-                ScalarValue::LargeBinary(a) => Ok(encoding
-                    .encode_large_scalar(a.as_ref().map(|v: &Vec<u8>| v.as_slice()))),
-                other => exec_err!(
-                    "Unsupported data type {other:?} for function encode({encoding})"
-                ),
-            }
+        ScalarValue::Binary(maybe_bytes) | ScalarValue::BinaryView(maybe_bytes) => {
+            Ok(ColumnarValue::Scalar(ScalarValue::Utf8(
+                maybe_bytes
+                    .as_ref()
+                    .map(|bytes| encoding.encode_bytes(bytes)),
+            )))
         }
-    }
-}
-
-fn decode_process(value: &ColumnarValue, encoding: Encoding) -> Result<ColumnarValue> {
-    match value {
-        ColumnarValue::Array(a) => match a.data_type() {
-            DataType::Utf8 => encoding.decode_utf8_array::<i32>(a.as_ref()),
-            DataType::LargeUtf8 => encoding.decode_utf8_array::<i64>(a.as_ref()),
-            DataType::Utf8View => encoding.decode_utf8_array::<i32>(a.as_ref()),
-            DataType::Binary => encoding.decode_binary_array::<i32>(a.as_ref()),
-            DataType::LargeBinary => encoding.decode_binary_array::<i64>(a.as_ref()),
-            other => exec_err!(
-                "Unsupported data type {other:?} for function decode({encoding})"
-            ),
-        },
-        ColumnarValue::Scalar(scalar) => {
-            match scalar {
-                ScalarValue::Utf8(a) => {
-                    encoding.decode_scalar(a.as_ref().map(|s: &String| s.as_bytes()))
-                }
-                ScalarValue::LargeUtf8(a) => encoding
-                    .decode_large_scalar(a.as_ref().map(|s: &String| s.as_bytes())),
-                ScalarValue::Utf8View(a) => {
-                    encoding.decode_scalar(a.as_ref().map(|s: &String| s.as_bytes()))
-                }
-                ScalarValue::Binary(a) => {
-                    encoding.decode_scalar(a.as_ref().map(|v: &Vec<u8>| v.as_slice()))
-                }
-                ScalarValue::LargeBinary(a) => encoding
-                    .decode_large_scalar(a.as_ref().map(|v: &Vec<u8>| v.as_slice())),
-                other => exec_err!(
-                    "Unsupported data type {other:?} for function decode({encoding})"
-                ),
-            }
+        ScalarValue::LargeBinary(maybe_bytes) => {
+            Ok(ColumnarValue::Scalar(ScalarValue::LargeUtf8(
+                maybe_bytes
+                    .as_ref()
+                    .map(|bytes| encoding.encode_bytes(bytes)),
+            )))
         }
+        v => internal_err!("Unexpected value for encode: {v}"),
     }
 }
 
-fn hex_encode(input: &[u8]) -> String {
-    hex::encode(input)
+fn encode_array(array: &ArrayRef, encoding: Encoding) -> Result<ColumnarValue> {
```

Comment:
These `encode_scalar`, `encode_array`, `decode_scalar`, `decode_array` methods should now be simpler as we've removed consideration of string types (expect signature to coerce them to binary for us)